src/request.rs

```diff
@@ -42,6 +43,144 @@ impl AppendOnlyData {
+            )),
+            AppendOnlyData::UnpubUnseq(data) => Ok(ADataIndices::new(
+                data.entry_index(),
+                data.owners_index(),
+                data.permissions_index(),
+            )),
+        }
+    }
+
+    pub fn last_entry(&self) -> Option<(Vec<u8>, Vec<u8>)> {
+        match self {
+            AppendOnlyData::PubSeq(data) => data.last(),
+            AppendOnlyData::PubUnseq(data) => data.last(),
+            AppendOnlyData::UnpubSeq(data) => data.last(),
+            AppendOnlyData::UnpubUnseq(data) => data.last(),
+        }
+    }
+
+    pub fn get_owners(&self, idx: u64) -> Option<&ADataOwner> {
+        match self {
+            AppendOnlyData::PubSeq(data) => data.fetch_owner_at_index(idx),
+            AppendOnlyData::PubUnseq(data) => data.fetch_owner_at_index(idx),
+            AppendOnlyData::UnpubSeq(data) => data.fetch_owner_at_index(idx),
+            AppendOnlyData::UnpubUnseq(data) => data.fetch_owner_at_index(idx),
+        }
+    }
+    pub fn get_pub_user_permissions(
+        &self,
+        user: ADataUser,
+        idx: u64,
+    ) -> Result<ADataPubPermissionSet, Error> {
+        match self {
+            AppendOnlyData::PubSeq(data) => match data.fetch_permissions_at_index(idx) {
+                Some(perm) => match perm.clone().permissions().get(&user) {
+                    Some(usr) => Ok(*usr),
+                    None => Err(Error::NoSuchEntry),
+                },
+                None => Err(Error::NoSuchEntry),
+            },
+            AppendOnlyData::PubUnseq(data) => match data.fetch_permissions_at_index(idx) {
+                Some(perm) => match perm.clone().permissions().get(&user) {
+                    Some(usr) => Ok(*usr),
+                    None => Err(Error::NoSuchEntry),
+                },
+                None => Err(Error::NoSuchEntry),
+            },
+            _ => Err(Error::NoSuchData),
+        }
+    }
+
+    pub fn get_unpub_user_permissions(
+        &self,
+        user: PublicKey,
+        idx: u64,
+    ) -> Result<ADataUnpubPermissionSet, Error> {
+        match self {
+            AppendOnlyData::UnpubSeq(data) => match data.fetch_permissions_at_index(idx) {
+                Some(perm) => match perm.clone().permissions().get(&user) {
+                    Some(usr) => Ok(*usr),
+                    None => Err(Error::NoSuchEntry),
+                },
+                None => Err(Error::NoSuchEntry),
+            },
+            AppendOnlyData::UnpubUnseq(data) => match data.fetch_permissions_at_index(idx) {
+                Some(perm) => match perm.clone().permissions().get(&user) {
+                    Some(usr) => Ok(*usr),
+                    None => Err(Error::NoSuchEntry),
+                },
+                None => Err(Error::NoSuchEntry),
+            },
+            _ => Err(Error::NoSuchData),
+        }
+    }
+
+    pub fn get_shell(&self, idx: u64) -> Result<Self, Error> {
+        match self {
```

Comment:
This could be:
```rust
use AppendOnlyData::*;
match self {
    PubSeq(adata) => adata.shell(idx).map(PubSeq),
    PubUnseq(adata) => adata.shell(idx).map(PubUnseq),
    UnpubSeq(adata) => adata.shell(idx).map(UnpubSeq),
    UnpubUnseq(adata) => adata.shell(idx).map(UnpubUnseq),
}
```